type errors

Bug Report

Mongoose 5.11.2 and nodejs 10.23






i have a schema that include this code:

.
.
.
    location: {
      type: { type: String },
      coordinates: [ Number ]
    },
.
.
.


when i try to build this, i receive this error:

error TS2322: Type '{ type: { type: StringConstructor; }; coordinates: NumberConstructor[]; }' is not assignable to type 'string | Function | Function[] | Schema | SchemaTypeOptions<any> | Schema[] | SchemaTypeOptions<any>[]'.
  Object literal may only specify known properties, and 'coordinates' does not exist in type 'Function | Function[] | Schema | SchemaTypeOptions<any> | Schema[] | SchemaTypeOptions<any>[]'.

16       coordinates: [ Number ]




it always worked using @types/mongoose, but since the last updates it is crashing on build.

--- a/index.d.ts
+++ b/index.d.ts
@@ -1047,7 +1047,7 @@ declare module "mongoose" {
   }
 
   interface SchemaDefinition {
-    [path: string]: SchemaTypeOptions<any> | Function | string | Schema | Schema[] | Array<SchemaTypeOptions<any>> | Function[];
+    [path: string]: SchemaTypeOptions<any> | Function | string | Schema | Schema[] | Array<SchemaTypeOptions<any>> | Function[] | SchemaDefinition | SchemaDefinition[];
   }
 
   interface SchemaOptions {

--- a/test/typescript/createBasicSchemaDefinition.ts
+++ b/test/typescript/createBasicSchemaDefinition.ts
@@ -10,7 +10,7 @@ const schema: Schema = new Schema({
 interface ITest extends Document {
   name?: string;
   tags?: string[];
-  authors?: { name: string };
+  author?: { name: string };
   followers?: { name: string }[];
 }
 

--- a/test/typescript/createBasicSchemaDefinition.ts
+++ b/test/typescript/createBasicSchemaDefinition.ts
@@ -2,12 +2,16 @@ import { Schema, model, Document } from 'mongoose';
 
 const schema: Schema = new Schema({
   name: { type: 'String' },
-  tags: [String]
+  tags: [String],
+  author: { name: String },
+  followers: [{ name: String }]
 }, { collection: 'mytest', versionKey: '_version' });
 
 interface ITest extends Document {
   name?: string;
   tags?: string[];
+  authors?: { name: string };
+  followers?: { name: string }[];
 }
 
 const Test = model<ITest>('Test', schema);

